Tutorial: highlight necessary content of the WORKSPACE file

smelc opened the pull request:
PR for issue #780 

Highlight that the way to obtain rules_haskell in tutorial/WORKSPACE is limited to the tutorial's case, otherwise better refer to the template created by start

also link to Common Haskell Build Use Cases to highlight the other mandatory thing in WORKSPACE file: obtaining GHC

I tried to make a minimal addition, so as to avoid polluting the tutorial; yet giving the sufficient pointers to make users able to set up a toolchain

Profpatsch commented on docs/haskell.rst:
```diff
@@ -89,14 +89,32 @@ special:
+
+File ``tutorial/WORKSPACE`` defines how to obtain ``rules_haskell``
+and a compiler toolchain.  Because this tutorial lives in ``rules_haskell``'s
+repository, ``rules_haskell`` is made available as follows::
```
```suggestion
repository, we reference ``rules_haskell`` as follows::
```

Profpatsch commented on docs/haskell.rst:
```diff
@@ -89,14 +89,32 @@ special:
+  )
+
+We refer to the template created by the start_ script
+to make ``rules_haskell`` available outside a local clone.
```
```suggestion
If you want to set up the `WORKSPACE` file for your own repository,
run the start_ script and it will generate a fitting configuration.
```

aherrmann commented:
> Maybe we should make it clearer in the tutorial, that users should not copy the `tutorial` files, but run the `start` script.

Agreed, and I'd go a step further and remove the `local_repository` snippet. A good guideline: If you show a code snippet, then it should be correct to copy-paste it.

Instead, you can have a sentence pointing out that `tutorial/WORKSPACE` only works within the `rules_haskell` repository. For user's own projects point to the `start` script, and maybe also include a `http_archive` snippet. Most rules sets have a copy-pasteable `http_archive` snippet in their setup instructions. E.g. [`rules_sh`](https://github.com/tweag/rules_sh#setup), [`rules_go`](https://github.com/bazelbuild/rules_go#setup), [`rules_nodejs`](https://bazelbuild.github.io/rules_nodejs/install.html).

Profpatsch commented:
> A good guideline: If you show a code snippet, then it should be correct to copy-paste it.

Haha, I had a sentence like this, but then deleted it again. So yeah, +1

Profpatsch commented:
> and maybe also include a `http_archive` snippet. Most rules sets have a copy-pasteable `http_archive` snippet in their setup instructions. E.g. [`rules_sh`](https://github.com/tweag/rules_sh#setup), [`rules_go`](https://github.com/bazelbuild/rules_go#setup), [`rules_nodejs`](https://bazelbuild.github.io/rules_nodejs/install.html).

Even if you get the `http_archive` stuff right, it’s still *way* too easy to miss some implicit invocation and have nothing work with abysmal error messages. I would put heavy focus on telling people to *always* use the start script.

smelc commented:
> Agreed, and I'd go a step further and remove the `local_repository` snippet. A good guideline: If you show a code snippet, then it should be correct to copy-paste it.

I agree, but the solutions we're discussing are working around the fact that the tutorial fetches `rules_haskell` in a way that doesn't work for production.

Is there a good reason for that? Could we change the tutorial it so that it fetches `rules_haskell` using `http_archive` like the `start` script does? We could still leave the `local_archive` example commented in `WORKSPACE`, as I understand it can be convenient to work against the local clone for experimenting.

smelc commented:
To proceed, in ea30ef6a3fd1714814cd40aff2deac9a648f036b, I added a `http_archive` snippet in the _use cases_ page and referenced it in the tutorial. I also removed the `local_repository` snippet I had added in my first commit in the tutorial.

aherrmann commented on docs/haskell.rst:
```diff
@@ -89,14 +89,27 @@ special:
+Understand the WORKSPACE file
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+File ``tutorial/WORKSPACE`` defines how to obtain ``rules_haskell``.
```
```suggestion
The file ``tutorial/WORKSPACE`` defines how to obtain ``rules_haskell``.
```

smelc replied:
Done in bec27b4.

aherrmann commented on docs/haskell.rst:
```diff
@@ -89,14 +89,27 @@ special:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+File ``tutorial/WORKSPACE`` defines how to obtain ``rules_haskell``.
+This file works within the ``rules_haskell``
```
```suggestion
This file only works within the ``rules_haskell``
```

smelc replied:
Done in bec27b4.
